CollectionView Android - Fix adapter update

andreinitescu opened the pull request:
### Description of Change ###

On Android, there's some issues with how the `ItemsViewRenderer.UpdateAdapter()` works:
1. It just resets to the `ItemsViewAdapter` instance but without taking into account if the empty view adapter is the current adapter.
2. `SelectableItemsViewRenderer` overrides `UpdateAdapter()` and it suffers from issue #1 too 
3. `SelectableItemsViewRenderer` overrides `UpdateAdapter()` it doesn't dispose the previous `ItemsViewAdapter` instance how the base class `ItemsViewRenderer` does in `UpdateAdapter()`

There's also an issue with how `ItemsViewRenderer.UpdateEmptyViewVisibility()` works.
I think it shouldn't reset the adapter and the layout manager without checking what's the current adapter, otherwise it creates issues, one is #6644

<!-- Describe your changes here. -->

### Issues Resolved ### 
<!-- Please use the format "fixes #xxxx" for each issue this PR addresses -->

- fixes #6644 (it resets the adapter and the layout manager hence stopping any animation)
- fixes #6839 

### API Changes ###

Added a new virtual method to ItemsViewRenderer:
```
protected virtual ItemsViewAdapter CreateAdapter()
{
    return new ItemsViewAdapter(ItemsView);
}
```

This is overridden by derived class `SelectableItemsViewRenderer` and this way makes the base class `UpdateAdapter()` work correctly with whatever the items adapter is.

### Platforms Affected ### 
<!-- Please list all platforms affected by these changes -->

- Android

### Behavioral/Visual Changes ###
<!-- Describe any changes that may change how a user's app behaves or appears when upgrading to this version of the codebase. -->

None

### Before/After Screenshots ### 
<!-- If possible, take a screenshot of your test case before these changes were made and another screenshot after the changes were made to show possible visual changes. -->

Not applicable

### Testing Procedure ###
<!-- Please list the steps that should be taken to properly test these changes on each relevant platform. If you were unable to test these changes yourself on any or all platforms, please let us know. Also, if you are able to attach a video of your test run, you will be our personal hero. -->

### PR Checklist ###

- [ ] Has automated tests <!-- (if tests are omitted or manual, state reason in description) -->
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

jfversluis commented on Xamarin.Forms.Platform.Android/CollectionView/ItemsViewRenderer.cs:
```diff
@@ -227,13 +227,19 @@ protected virtual void UpdateItemsSource()
+			return new ItemsViewAdapter(ItemsView);
+		}
+
+		void UpdateAdapter()
```
Changing this one might be breaking for custom renderers. But maybe that’s OK since it’s still preview? @samhouts

andreinitescu replied:
Yes, since it's in preview, this should be OK from that perspective

samhouts replied:
Yep, we'll just add the `breaking` tag to give users a heads-up in the release notes. Thanks!

jfversluis commented on Xamarin.Forms.Platform.Android/CollectionView/SelectableItemsViewRenderer.cs:
```diff
@@ -37,10 +37,9 @@ protected override void SetUpNewElement(ItemsView newElement)
 			UpdateNativeSelection();
 		}
 
-		protected override void UpdateAdapter()
```
See my comment for the other one :)

Otherwise looks good to me

jfversluis commented:
Failing test seem unrelated

hartez commented:
@andreinitescu This needs to be retargeted to 4.2.0. Also, #6879 made some changes to how Android handles EmptyView; I don't know if that will affect this PR or now.

andreinitescu commented:
My PR has been submitted 6 days ago. You submitted yours yesterday. Could you please take a look and tell if this PR still applies. Thank you.

hartez commented:
Looks like the changes from #6879 _do_ fix #6839, but they _do not_ fix #6644.

hartez commented:
@andreinitescu Also, apologies for the PR clash; the original version of #6879 (#6172) was finished in May, but got stuck in rebase/UI test hell for a long time.

andreinitescu commented:
@hartez No need to apologize, I just needed a bit of help to be sure I don't mess it up. 
I rebased to 4.2.0, and I added a gallery page for the issue #6644. Let me know in the case there's anything more needed.

jfversluis commented:
Added `#if APP` compiler directive to fix build

samhouts commented:
Failing tests not related
